fix(retention): fix breakdown people urls

hazzadous opened the pull request:
## Changes

This change returns people_url for each breakdown cohort in the
response. We also merge the initial and returning queries together,
as this makes it easier to align the people query also.
    
Note that I'm talking about person_id as opposed to actor_type etc.
but perhaps that can be a followup.

There are significant conflicts with master that I need to resolve first

## How did you test this code?

<!-- If the answer is manually, please include a quick step-by-step on how to test this PR. -->

*Please describe.*

Commit:
fix(retention): fix breakdown people urls

This change returns people_url for each breakdown cohort in the
response. We also merge the initial and returning queries together,
as this makes it easier to align the people query also.

Note that I'm talking about person_id as opposed to actor_type etc.
but perhaps that can be a followup.

Commit:
clean up clickhouse params

Commit:
tidy up a little

Commit:
remove import

Commit:
remove non-breakdown specific code

Commit:
make cohort by initial event date a special breakdown case

Commit:
keep date for backwards compat

hazzadous commented:
I still need to resolve conflicts fyi

Commit:
Remove unused sql

Commit:
make test stable

EDsCODE commented on ee/clickhouse/queries/clickhouse_retention.py:
```diff
@@ -1,181 +1,78 @@
-
-        return result_dict
+    def run(self, filter: RetentionFilter, team: Team, *args, **kwargs) -> List[Dict[str, Any]]:
+        retention_by_breakdown = self._get_retention_by_breakdown_values(filter, team)
```
This naming is confusing because it appears as if this function will only return breakdown based retention

EDsCODE commented on ee/clickhouse/queries/clickhouse_retention.py:
```diff
@@ -193,151 +99,120 @@ def process_breakdown_table_result(
         return result
+
+
+def build_actor_query(
```
another naming issue. once you merge and resolve conflicts with the clickhous retention actor class, it will be confusing to differentiate this actor query from the other

EDsCODE commented on ee/clickhouse/views/test/test_clickhouse_retention.py:
```diff
@@ -172,16 +178,77 @@ def test_can_specify_breakdown_event_property(self):
+        }
+
+    @test_with_materialized_columns(event_properties=["os"])
+    def test_can_specify_breakdown_event_property_and_retrieve_people(self):
```
maybe start snapshotting retention queries too

EDsCODE commented on ee/clickhouse/views/test/test_clickhouse_retention.py:
```diff
@@ -49,13 +49,16 @@ def test_can_get_retention_cohort_breakdown(self):
             ),
```
Tedious but worth moving and converting ee/clickhouse/queries/test/test_retention.py tests here and use the new helpers so we can gradually unify the tests. (Can also do this as a follow up)

hazzadous replied:
I think work leaving this to a separate pr? 🤞

EDsCODE reviewed (commented):
Clean up looks great. I will comb through the new query logic after the merge conflicts are resolved!

Commit:
wip

Commit:
Get most of the tests working

Commit:
test(retention): remove graph retention test

We no longer need this, we have all the information we need from the
table response for retention, and can construct this on the frontend.

Commit:
revert any changes to posthog/queries/retention.py

Commit:
revert any changes to ee/clickhouse/models/person.py

Commit:
Revert posthog/queries/retention.py to merge-base

Commit:
Merge remote-tracking branch 'origin/master' into
feat/fix-retention-people-urls

This is quite large, as needed to resolve actor change conflicts.

Commit:
Ensure actor id is a str

Commit:
Add type for actor serialiser for type narrowing

Commit:
run black

Commit:
sort imports

Commit:
Remove retention_actors.py

Commit:
fix typings

Commit:
format

EDsCODE committed:
reverse str type

EDsCODE committed:
sort imports

EDsCODE committed:
rename

EDsCODE committed:
split out functions

EDsCODE committed:
remove deuplicate logic

EDsCODE committed:
working

EDsCODE committed:
fix type

EDsCODE committed:
don't stringify

EDsCODE committed:
fix test

EDsCODE committed:
ordering doesn't matter

EDsCODE commented on ee/clickhouse/queries/retention/retention_actors.py:
```diff
@@ -57,70 +26,18 @@ def is_aggregating_by_groups(self) -> bool:
 
 
+# Note: This class does not respect the entire flor from ActorBaseQuery because the result shape differs from other actor queries
 class ClickhouseRetentionActorsByPeriod(ActorBaseQuery):
```
Just brought over a lot of the functions that were being used to create this functionality

hazzadous commented on ee/clickhouse/queries/retention/retention_actors.py:
```diff
@@ -94,7 +94,7 @@ def actors(self):
                 ],
             }
-            for actor in sorted(actor_appearances, key=lambda x: x.actor_id)
+            for actor in sorted(actor_appearances, key=lambda x: x.appearance_count, reverse=True)
```
non-blocking

fyi this won't work with pagination, i.e. the person with the highest count could be on page 2 I think

Commit:
trigger ci

hazzadous commented:
QA video: https://www.loom.com/share/37a100d736fd4a98af8693edb057c487
